### PR TITLE
Implement various AliasAnalysis operations directly on top of MemoryLocations.

### PR DESCRIPTION
**Pull request by resistor**
This reduces DenseNet load time by about 25% (down to 5.3s on my laptop) and gets AliasAnalysis out of the profile top hits entirely.

### CODE REVIEW

**Comment by resistor** [comment]
@Chillee

File: torch/csrc/jit/passes/utils/memory_dag.h [hunk 1]
```diff
@@ -126,9 +99,8 @@ struct Element {
   MemoryLocations pointedFrom;
 
   MemoryLocations contained_elements;
-  static unsigned indexCount;
-  signed index;
-  Element(const Value* value_);
+  int index;
```
Comment (Chillee) [inline torch/csrc/jit/passes/utils/memory_dag.h]: Do we want to make these unsigned?

**Review by suo — APPROVED** [review]
As discussed, we need to clean up the boundary between MemoryDAG and AliasDB. Out of scope for this PR though.

File: torch/csrc/jit/passes/alias_analysis.h [hunk 2]
```diff
@@ -78,21 +78,37 @@ class AliasDb {
...
+    for (auto it = a.cbegin(); it != a.cend();) {
+      const auto value = *it;
+      if (elementMap_.count(value)) {
+        auto element = elementMap_.at(value);
+        memoryLocations |= element->getMemoryLocations();
       }
+
+      do {
+        ++it;
+      } while (it != a.cend() && *it == value);
```
Comment (Chillee) [inline torch/csrc/jit/passes/alias_analysis.h]: Is it true that for an arbitrary unordered multiset, elements with the same key will be located contiguously when you iterate?

File: torch/csrc/jit/passes/alias_analysis.h [hunk 3]
```diff
@@ -78,21 +78,37 @@ class AliasDb {
...
+        if (memoryLocations.intersects(element->getMemoryLocations())) {
+          return true;
+        }
       }
-    }
 
-    return memoryDAG_->mayAlias(aElements, bElements);
+      do {
+        ++it;
+      } while (it != a.cend() && *it == value);
```
Comment (Chillee) [inline torch/csrc/jit/passes/alias_analysis.h]: Shouldn't this be `it != b.cend()`?

**Comment by Chillee** [comment]
Other than the comments, LGTM.

File: torch/csrc/jit/passes/alias_analysis.h (hunk 2 above)
Comment (resistor) [reply torch/csrc/jit/passes/alias_analysis.h]: @suo Thoughts here?  Doing it this way was measurably faster, and I'm not clear on what the actual multiset requirements are here.

Comment (resistor) [reply torch/csrc/jit/passes/alias_analysis.h]: That said, I don't *think* this differs from the original implementation.  Using std::advance also assumed they were contiguous?

Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.h]: yes

Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.h]: to both questions haha

Comment (Chillee) [reply torch/csrc/jit/passes/alias_analysis.h]: Ah that's true that it's the same API as before. LGTM then.

**Pull request closed without merging**